src/vmm/src/builder.rs

```diff
@@ -148,26 +148,344 @@ impl std::convert::From<linux_loader::cmdline::Error> for StartMicrovmError {
+        use self::StartMicrovmError::*;
+
+        // Construct the base CpuConfiguration to apply CPU template onto.
+        let cpu_config = {
+            use crate::arch::aarch64::regs::Aarch64RegisterVec;
+            use crate::arch::aarch64::vcpu::get_registers;
+
+            for vcpu in vcpus.iter_mut() {
+                vcpu.kvm_vcpu
+                    .init(&cpu_template.vcpu_features)
+                    .map_err(VmmError::VcpuInit)
+                    .map_err(Internal)?;
+            }
+
+            let mut regs = Aarch64RegisterVec::default();
+            get_registers(&vcpus[0].kvm_vcpu.fd, &cpu_template.reg_list(), &mut regs)
+                .map_err(GuestConfigError)?;
+            CpuConfiguration { regs }
+        };
+
+        // Apply CPU template to the base CpuConfiguration.
+        let cpu_config = CpuConfiguration::apply_template(cpu_config, cpu_template)?;
+
+        let vcpu_config = VcpuConfig {
+            vcpu_count: vm_config.vcpu_count,
+            // smt does not exist on aarch64
+            smt: false,
+            cpu_config,
+        };
+
+        // Configure vCPUs with normalizing and setting the generated CPU configuration.
+        for vcpu in vcpus.iter_mut() {
+            vcpu.kvm_vcpu
+                .configure(vmm.guest_memory(), entry_addr, &vcpu_config)
+                .map_err(VmmError::VcpuConfigure)
+                .map_err(Internal)?;
+        }
+
+        let vcpu_mpidr = vcpus
+            .iter_mut()
+            .map(|cpu| cpu.kvm_vcpu.get_mpidr())
+            .collect();
+        let cmdline = boot_cmdline.as_cstring()?;
+        crate::arch::aarch64::configure_system(
+            &vmm.guest_memory,
+            cmdline,
+            vcpu_mpidr,
+            vmm.mmio_device_manager.get_device_info(),
+            vmm.vm.get_irqchip(),
+            &vmm.acpi_device_manager.vmgenid,
+            initrd,
+        )
+        .map_err(ConfigureSystem)?;
+
+        Ok(())
+    }
+
+    /// Sets up the irqchip for a aarch64 microVM.
+    pub fn setup_interrupt_controller(vm: &mut Vm, vcpu_count: u8) -> Result<(), VmmError> {
+        vm.setup_irqchip(vcpu_count).map_err(VmmError::Vm)
+    }
+
+    /// On aarch64, the vCPUs need to be created (i.e call KVM_CREATE_VCPU) before setting up the
+    /// IRQ chip because the `KVM_CREATE_VCPU` ioctl will return error if the IRQCHIP
+    /// was already initialized.
+    /// Search for `kvm_arch_vcpu_create` in arch/arm/kvm/arm.c.
+    pub fn create_vcpus(
+        vm: &mut Vm,
+        vcpu_count: u8,
+        exit_evt: &EventFd,
+    ) -> Result<Vec<Vcpu>, VmmError> {
+        let mut vcpus = Vec::with_capacity(vcpu_count as usize);
+        for cpu_idx in 0..vcpu_count {
+            let exit_evt = exit_evt.try_clone().map_err(VmmError::EventFd)?;
+            let vcpu = Vcpu::new(cpu_idx, vm, exit_evt).map_err(VmmError::VcpuCreate)?;
+            vcpus.push(vcpu);
+        }
+
+        setup_interrupt_controller(vm, vcpu_count)?;
```

Comment:
Let's move creation of interrupt controller out of `create_vcpus` and into the `create_vmm_and_vcpus`.